oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java

```diff
@@ -78,26 +81,43 @@ public ServerToNodeMetadata(Function<Hdd, Volume> fnVolume,
    @Override
    public NodeMetadata apply(final Server server) {
       checkNotNull(server, "Null server");
+      String hardwareId = null;
 
       DataCenter dataCenter = api.dataCenterApi().get(server.datacenter().id());
       Location location = find(locations.get(), idEquals(dataCenter.id()));
       Hardware hardware = null;
+      Float size = null;
+      Integer minRam = 0;
+      List<Hdd> hdds = null;
+      List<Volume> volumes = Lists.newArrayList();
+      List<Processor> processors = new ArrayList<Processor>();
+
       //check if the server was built on a hardware flavour(Fixed instance)
-      if (server.hardware().fixedInstanceSizeId() != null && !"0".equals(server.hardware().fixedInstanceSizeId())) {
-         hardware = hardwareFlavors.get().get(server.hardware().fixedInstanceSizeId());
+      if (server.serverType() == Types.ServerType.baremetal) {
+         BareMetalModel bmModel = api.serverApi().getBaremetalModel(server.hardware().baremetalModelId());
+         hardware = hardwareFlavors.get().get(bmModel.name());
+         hardwareId = hardware.getId();
+
+      } else if (server.hardware().fixedInstanceSizeId() != null && !"0".equals(server.hardware().fixedInstanceSizeId())) {
+         HardwareFlavour flavor = api.serverApi().getHardwareFlavour(server.hardware().fixedInstanceSizeId());
+         hardware = hardwareFlavors.get().get(flavor.id() + "," + flavor.name());
+         hardwareId = hardware.getId();
 
       } else {
-         List<Volume> volumes = Lists.newArrayList();
          //customer hardware
-         double size = 0d;
-         double minRam = server.hardware().ram();
-         List<Hdd> hdds = server.hardware().hdds();
+         size = 0F;
+         minRam = (int) server.hardware().ram();
+         hdds = server.hardware().hdds();
 
          if (server.hardware().hdds().isEmpty()) {
             hdds = api.serverApi().getHardware(server.id()).hdds();
          }
 
-         size = getHddSize(hdds);
+         size = (float) getHddSize(hdds);
+         if (size == 0) {
+            Server refreshedServer = api.serverApi().get(server.id());
```

Comment:
you should avoid the extra `api` calls inside `ServerToNodeMetadata` - please remove `api` injection and refactor this class to not use it